### DISCUSSION TITLE
How to configure colors with CSS variables (including opacity)?

### DISCUSSION THREAD

**Discussion by lockinmarv** [unocss/unocss#732] [Q&A] [answered]
Hey! How can I configure colors with CSS variables including the opacity like tailwind describes it here https://tailwindcss.com/docs/customizing-colors#using-css-variables?

Defining the color as a function like `primary: () => 'blue'` doesn't seem to work either.

Am I missing something or is this just not possible?

**Comment by zyyv** [comment]


**Reply by lockinmarv** [reply]
Thanks, can I use it as class instead of attribute because it doesn't work for me. It doesn't convert `#f00` to the correct rgb color. So it ends up with `rgba(#f00, 1)`. Also applying `text-opacity-50` doesn't work, the alpha property is always 1.

**Comment by chu121su12** [comment, marked as answer]
Something [like this](https://unocss.antfu.me/?html=FAHgJglgbgBAxgGwIYGcUF4BEAnA9rgF3gFo5YAjAc1KgHoBGABkwD456RbIoWg&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmcAvnOlBCHAOQCuWEAxgM6u0BQnqAHpLBQYAhvQA28NJhz5CwIgAoEnOHFYALaDGb0YrAFxwA2itXHaVCDFoAaOq1RjUzGNAC0R-ZZgBdfUbc3ADdhKCxhEFR9ACYABlibBNifWh9THxtTGHVUSMNlMzhmCDFoA0RTQuYgwwsiACMFEKgFQObwyIBKTttKskzVUkzSTqA&options=N4IgzgLgTglgxhEAuaBXApgXyA)?

To use the opacity utility separately, use base color syntax, not the color/opacity shorthand.

rgba & hsla needs comma separated color value. other color needs to be in space separated. ([hwb example](https://unocss.antfu.me/?html=FAHgJglgbgBAxgGwIYGcUF4BEAnA9rgF3gFo5YAjAc1KgHoBGABkwD456RbIoWg&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmcAvnOlBCHAOQCuWEAxgM6u0BQnqAHpLBQYAhvQA28NJhz5CwIgAoEnOHFYALaDGb0YrAFxwA2itXHaVCDFoAaOq1RjUzGNAC0R-ZZgBdfUbc3ADdhKCxhEFR9AE4ABgB9AEZYgFIk1J9aH1MfG1MYdVRIw2UzOGYIMWgDRFMy5iDDWnUAdwAjBRCoBUCu8MiASgHbOrI81VI80gGgA&options=N4IgzgLgTglgxhEAuaBXApgXyA))

**Reply by lockinmarv** [reply]
Thanks! It works indeed in a "normal" setup but I'm using the shadow-dom mode + `@unocss-placeholder` and it looks like it doesn't work there. The CSS that is generated isn't escaped and just printed in the <style> as `bg-primary/50`. Your example compiles the CSS to `bg-primary\/50`. Manually escaping it inside the devtools works.

**Reply by chu121su12** [reply]
Please file an issue regarding the shadow dom. For now, try setting the opacity in the color definition
[playground](https://unocss.antfu.me/?html=FAHgJglgbgBAxgGwIYGcUF4BEAnA9rgF2BhNLJLgFo4pjz6BtSylXAWwFNLcAHALgCcABgC6MJpShJsAOySc%2BARiEB9AExCApCoDMAFk0i69MpgB8cRSAD0kKGaA&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmcAvnOlBCHAOQCuWEAxgM6u0BQnqAHpLBQYAhvQA28NJhz5CwIgAoEnOHFYALaDGb0YrAFxwA2itXHaVCDFoAaOAANTZs61RjUzGNAC0R-ZZgAXX0FJ2czI29vADdhKCxhEFR9AE4ABgB9AEY0gFJsvMCw8OMo1mpUbwgwfQAmQuKzAEow%2ByLVQJtTGHVUJMNlZ2YIMWgDREbmaMNadQB3ACMFWKgFKJWEpKaAehW173KkqrA7LKam2zDSLtVrzlImoA&options=N4IgzgLgTglgxhEAuaBXApgXyA)

**Reply by lockinmarv** [reply]
For reference #739

**Comment by leomp12** [comment]
Hey!
`.bg-primary/20` with `--c-primary: rgb(1, 1, 1)` (for example) used to work but is not working anymore 😅
It's a bug or should I define variables another way? If it's a bug I can create an issue with details.

**Comment by Yves852** [comment]
I was looking for a solution aswell. Trying to figure out I noticed that, either by Vite (vite.config.ts) or Nuxt (uno.config.ts) colors defined directly in the config file will work with the shorthand `text-primary/50` but defining css variables and calling it in the config file will simply ignore the transparency shorthand.

Here a stackblitz repro. Notice the branch "nuxt" to compare both configs and commits for both cases.
https://stackblitz.com/~/github.com/Yves852/unocss-unocss-io46kp-transparency

My motivation here was to define an unique source of truth for color variables with light/dark themes in a easy way. 

**Edit**: managed it by converting hex colors to rgb. Sorry for the noise. Updated the stackblitz.

**Comment by AlexRMU** [comment]
Here is my option for opacity
```ts
    shortcuts: [
        [
            /^([A-Za-z0-9-:@]+)-\[(.+?)\]\/(\d{1,3})$/,
            ([, prefix, raw, alpha_str]) => {
                if (raw.includes("/")) {
                    return;
                }
                const alpha = Math.max(0, Math.min(100, parseInt(alpha_str) || 0));
                if (alpha === 0) {
                    return `${prefix}-transparent`;
                }
                let value = raw.replace(/_/g, " ");
                if (value.startsWith("--")) {
                    value = `var(${value})`;
                }
                const color_mix = `color-mix(in srgb, ${value} ${alpha}%, transparent)`;
                const escaped = color_mix.replace(/\s+/g, "_");
                return `${prefix}-[${escaped}]`;
            },
        ],
    ],
```
